Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -978,143 +963,202 @@ extension SwiftLanguageServer {
+            start: node.statements.position.utf8Offset,
+            end: node.rightBrace.positionAfterSkippingLeadingTrivia.utf8Offset)
         }
 
-        // Add the last stored comment.
-        if let comment = currentComment, hasReachedLimit == false {
-          self.addFoldingRange(offset: comment.offset, length: comment.length, kind: .comment, in: snapshot, toSet: &ranges)
-          currentComment = nil
+        override func visit(_ node: AccessorBlockSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.accessors.position.utf8Offset,
+            end: node.rightBrace.positionAfterSkippingLeadingTrivia.utf8Offset)
         }
 
-        var structureStack: [SKDResponseArray] = [substructure]
-        while !hasReachedLimit, let substructure = structureStack.popLast() {
-          substructure.forEach { _, value in
-            if let offset: Int = value[self.keys.bodyoffset],
-               let length: Int = value[self.keys.bodylength],
-               length > 0
-            {
-              self.addFoldingRange(offset: offset, length: length, in: snapshot, toSet: &ranges)
-              if hasReachedLimit {
-                return false
-              }
-            }
-            if let substructure: SKDResponseArray = value[self.keys.substructure] {
-              structureStack.append(substructure)
+        override func visit(_ node: SwitchStmtSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.cases.position.utf8Offset,
+            end: node.rightBrace.positionAfterSkippingLeadingTrivia.utf8Offset)
+        }
+
+        override func visit(_ node: FunctionCallExprSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.argumentList.position.utf8Offset,
+            end: node.argumentList.endPosition.utf8Offset)
+        }
+
+        override func visit(_ node: SubscriptExprSyntax) -> SyntaxVisitorContinueKind {
+          return self.addFoldingRange(
+            start: node.argumentList.position.utf8Offset,
+            end: node.argumentList.endPosition.utf8Offset)
+        }
+
+        __consuming func finalize() -> Set<FoldingRange> {
+          return self.ranges
+        }
+
+        private func addFoldingRange(start: Int, end: Int, kind: FoldingRangeKind? = nil) -> SyntaxVisitorContinueKind {
+          if let limit = self.rangeLimit, self.ranges.count >= limit {
+            return .visitChildren
+          }
+
+          guard let start: Position = snapshot.positionOf(utf8Offset: start),
+                let end: Position = snapshot.positionOf(utf8Offset: end) else {
+            log("folding range failed to retrieve position of \(snapshot.document.uri): \(start)-\(end)", level: .warning)
+            return .visitChildren
+          }
+          let range: FoldingRange
+          // If the client only supports folding full lines, ignore the end character's line.
```

Comment:
You probably meant the following, right?
```suggestion
          // If the client only supports folding full lines, ignore the end character's correct.
```

Reply:
I've updated the doc comment to be clearer. This was the old comment FWIW.